rhp/v4/rpc.go

```diff
@@ -0,0 +1,718 @@
+	} else if n > 0 {
+		// add change output
+		formationTxn.SiacoinOutputs = append(formationTxn.SiacoinOutputs, types.SiacoinOutput{
+			Address: fc.HostOutput.Address,
+			Value:   hostInputSum.Sub(fc.TotalCollateral),
+		})
+	}
+
+	// sign the renter inputs
+	signer.SignV2Inputs(&formationTxn, toSign)
+	formationSigHash := cs.ContractSigHash(fc)
+	formationTxn.FileContracts[0].RenterSignature = signer.SignHash(formationSigHash)
+
+	renterPolicyResp := rhp4.RPCFormContractSecondResponse{
+		RenterContractSignature: formationTxn.FileContracts[0].RenterSignature,
+	}
+	for _, si := range formationTxn.SiacoinInputs[:len(renterSiacoinElements)] {
+		renterPolicyResp.RenterSatisfiedPolicies = append(renterPolicyResp.RenterSatisfiedPolicies, si.SatisfiedPolicy)
+	}
+	// send the renter signatures
+	if err := rhp4.WriteResponse(s, &renterPolicyResp); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to write signature response: %w", err)
+	}
+
+	// read the finalized transaction set
+	var hostTransactionSetResp rhp4.RPCFormContractThirdResponse
+	if err := rhp4.ReadResponse(s, &hostTransactionSetResp); err != nil {
+		return RPCFormContractResult{}, fmt.Errorf("failed to read final response: %w", err)
+	}
+
+	if len(hostTransactionSetResp.TransactionSet) == 0 {
+		return RPCFormContractResult{}, fmt.Errorf("expected at least one host transaction")
+	}
+	hostFormationTxn := hostTransactionSetResp.TransactionSet[len(hostTransactionSetResp.TransactionSet)-1]
+	if len(hostFormationTxn.FileContracts) != 1 {
+		return RPCFormContractResult{}, fmt.Errorf("expected exactly one contract")
+	}
+
+	// check for no funny business
+	formationTxnID := formationTxn.ID()
+	hostFormationTxnID := hostFormationTxn.ID()
+	if formationTxnID != hostFormationTxnID {
+		return RPCFormContractResult{}, fmt.Errorf("expected transaction IDs to match %v != %v", formationTxnID, hostFormationTxnID)
+	}
+
+	// validate the host signature
+	fc.HostSignature = hostFormationTxn.FileContracts[0].HostSignature
+	if !fc.HostPublicKey.VerifyHash(formationSigHash, fc.HostSignature) {
+		return RPCFormContractResult{}, errors.New("invalid host signature")
+	}
+
+	return RPCFormContractResult{
+		Contract: ContractRevision{
+			ID:       formationTxn.V2FileContractID(formationTxnID, 0),
+			Revision: fc,
+		},
+		FormationSet: TransactionSet{
+			Basis:        hostTransactionSetResp.Basis,
+			Transactions: hostTransactionSetResp.TransactionSet,
+		},
+		Cost: renterCost,
+	}, nil
+}
+
+// RPCRenewContract renews a contract with a host
+func RPCRenewContract(ctx context.Context, t TransportClient, cr ChainReader, signer FormContractSigner, p rhp4.HostPrices, existing types.V2FileContract, params rhp4.RPCRenewContractParams) (RPCRenewContractResult, error) {
+	renewal := rhp4.NewRenewal(existing, p, params)
+	renewalTxn := types.V2Transaction{
+		MinerFee: types.Siacoins(1),
```

Comment:
:eyes:

Reply:
Is this about the miner fee? It's totally about the miner fee isn't it?

Reply:
it's about the miner fee

honestly 1 SC is a fine miner fee (well above average, probably) but hardcoding it...

Reply:
Yeah, we should not do that 😅